### PR TITLE
implement the full set of sorts on QueryManyIter

### PR DESCRIPTION
**Pull request by Victoronz**
# Objective

~Blocked on #13417~

Motivation is the same as in #13417. If users can sort `QueryIter`, to only makes sense to also allow them to use this functionality on `QueryManyIter`. 

## Solution

Also implement the sorts on `QueryManyIter`. 

The implementation of the sorts themselves are mostly the same as with `QueryIter` in #13417.
They differ in that they re-use the `entity_iter` passed to the `iter_many`, and internally call `iter_many_unchecked_manual` on the lens `QueryState` with it.
These methods also return a different struct, `QuerySortedManyIter`, because there is no longer a guarantee of unique entities.
`QuerySortedManyIter` implements the various `Iterator` traits for read-only iteration, as `QueryManyIter` does + `DoubleEndedIterator`.
For mutable iteration, there is both a `fetch_next` and a `fetch_next_back` method. However, they only become available after the user calls `collect_inner` on `QuerySortedManyIter` first. This collects the inner `entity_iter` (this is the sorted one, **not** the original the user passed) to drop all query lens items to avoid aliasing. 
When TAITs are available this `collect_inner` could be hidden away, until then it is unfortunately not possible to elide this without either regressing read-only iteration, or introducing a whole new type, mostly being a copy of `QuerySortedIter`.

As a follow-up we could add a `entities_all_unique` method to check whether the entity list consists of only unique entities, and then return a `QuerySortedIter` from it (under opaque impl Trait if need be), *allowing mutable `Iterator` trait iteration* over what was originally an `iter_many` call.
Such a method can also be added to `QueryManyIter`, albeit needing a separate, new return type.

## Testing

I've switched the third example/doc test under `sort` out for one that shows the collect_inner/fetch_next_back functionality, otherwise the examples are the same as in #13417, adjusted to use `iter_many` instead of `iter`.

The `query-iter-many-sorts` test checks for equivalence to the underlying sorts.
The test after shows that these sorts *do not* panic after `fetch`/`fetch_next` calls.

## Changelog

Added `sort`, `sort_unstable`, `sort_by`, `sort_unstable_by`, `sort_by_key`, `sort_by_cached_key` to `QueryManyIter`.
Added `QuerySortedManyIter`.

### CODE REVIEW

**Comment by Victoronz** [comment]
This currently doesn't include the #13417 commit, aside from the `NeutralOrd` helper, they are independent.

**Review by alice-i-cecile — APPROVED** [review]
These should exist, the docs are excellent, the engineering strategy is solid and well-justified and there's a good set of tests. Nice work.

**Review by chescock — APPROVED** [review]
This looks good!  I wrote some musings on possible ways to simplify things, but none of them should block this PR.

File: crates/bevy_ecs/src/query/iter.rs
Comment (Victoronz) [reply crates/bevy_ecs/src/query/iter.rs:2199]: Let me rephrase for clarity:
In an iteration with no duplicates (`QuerySortedIter`) this presents no issue because the offending reference is always dropped before `D::Item` is retrieved.
However with duplicates, a mutable reference to `D::Item` could alias with a not yet dropped `L::Item` later on in the iteration, which means all `L::Item`s need to be dropped before we allow mutable iteration via `fetch_next`.

`ManuallyDrop` does not make sense here, since you either want to drop nothing in advance (lazy map) or everything (eager map).
The problem here is that iterators are lazy, and especially when sorting, you may just want the first or last item.
Paying for an eager map of the entire query seemed wasteful, so I implemented `collect_inner`.
This problem stems from mutable and immutable iteration returning the same struct. 

Maybe we could actually have two impl blocks for `QueryManyIter`, one for mutable `QueryData`, and one for immutable. It would be nice if you could have sort methods with the same name on both, returning an eagerly mapped or lazily mapped iterator respectively.

File: crates/bevy_ecs/src/query/iter.rs
Comment (Victoronz) [reply crates/bevy_ecs/src/query/iter.rs:2653]: You got it! More documentation should definitely be made. Currently, the docs for `Query` do not mention sorting, which they really should as well.

Yeah, there are a lot of iterator types, however it is normal in the rust ecosystem to have different structs for different forms of iteration, just like every adapter on `iterator` returns its own struct. (That bevy has combined mutable and immutable into a single type is already unusual)

However I do not think combining them is a good idea. That would not reduce code duplication by much, since different forms of iteration will still need their own impl blocks regardless.
Quite the opposite, I would find it more confusing both from a dev, and the user standpoint.
Since picking the "wrong kind" of iteration is quite common, and iterators like to nest, it is very important to be as clear as we can inside errors. (Errors unpack type aliases)
Besides, future iterator types that actually do just do the same kind of iteration can be type aliases in clear manner:
`pub type QuerySortedManyUniqueIter<...> = QuerySortedIter<...>;`

File: crates/bevy_ecs/src/query/iter.rs
Comment (Victoronz) [reply crates/bevy_ecs/src/query/iter.rs:1484]: I had tried, but ran into strange lifetime issues, so I punted on it. I also wanted to preserve when `Entity` was wrapped in `NeutralOrd`.

File: crates/bevy_ecs/src/query/iter.rs
Comment (chescock) [reply crates/bevy_ecs/src/query/iter.rs:2199]: Okay, it sounds like I understand the aliasing issue!  

Regarding eager vs. lazy iteration: Note that you still have to drop all of the `L::Item`s even if you just take the first or last item.  It just happens implicitly when you drop the `Map<IntoIter>` without fully iterating it.  

(And, of course, most *actual* `L::Item` types like `&T` and `Mut<T>` have no drop glue.  So we're doing nothing, and the question is *when* we're doing nothing :).)  

The expensive thing we want to avoid in the immutable case is having to allocate a fresh `Vec<Entity>` and copy everything over.  And it would be even better if we could avoid it in the mutable case, too!  

I think something like this works: 
```rust
let mut keyed_query: Vec<_> = query_lens
    .map(|(key, entity)| (ManuallyDrop::new(key), NeutralOrd(entity)))
    .collect();
keyed_query.sort();
for (item, _) in &mut keyed_query {
    // SAFETY: `keyed_query` is immediately consumed by the `map`, which ignores the item,
    // so the value is never exposed to safe code or dropped again.
    unsafe { ManuallyDrop::drop(item) };
}
let entity_iter = keyed_query.into_iter().map(|(.., entity)| entity.0);
```

We're eagerly dropping the lens items, but we have to drop those eventually, so that's moving work around rather than adding anything.  And for the common case where there is no drop glue, the compiler can eliminate the entire loop.  

There is no extra `collect()`, so there is no extra memory allocation or copying.  

And the lens items have all been dropped, so there are no longer aliasing concerns and we can offer `fetch_next` to mutable queries!

File: crates/bevy_ecs/src/query/iter.rs
Comment (chescock) [reply crates/bevy_ecs/src/query/iter.rs:2653]: My concern with duplication is mainly the various `fetch_next` methods, which duplicate the logic of `Query::get`.  `QueryManyIter::fetch_next_aliased_unchecked`, `QuerySortedIter::fetch_next`, and `QuerySortedManyIter::fetch_next_aliased_unchecked` are almost identical, and it's hard to see what the differences are if you want to make changes to the logic.  

If we want to have multiple public types, then maybe we can have a single inner type with the complicated `fetch_next` implementation, and have the other types wrap that with appropriate safety assertions.  

But we can always refactor once this and #13477 are both merged, so I don't think we need to worry about it yet!

File: crates/bevy_ecs/src/query/iter.rs
Comment (Victoronz) [reply crates/bevy_ecs/src/query/iter.rs:2199]: I don't think this will compile/is safe.
See this example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=df1f5c8daf4f12206f4e954701efc3b7

Calling the destructor of a reference will not end its borrow, since that doesn't deallocate!

File: crates/bevy_ecs/src/query/iter.rs
Comment (Victoronz) [reply crates/bevy_ecs/src/query/iter.rs:2653]: There are several properties about their entities query iterators can rely on/guarantee:

Iterator | Uniqueness | Existence | QF eval* | dense EL/A order**
--- | --- | --- | --- | --- 
QueryManyIter | Unknown | Unknown | Unknown | Unknown
QueryManyUniqueIter | Known | Unknown | Unknown | Unknown
QuerySortedManyIter | Unknown | Known | Known | Depends
QuerySortedManyUniqueIter | Known | Known | Known | Depends
QuerySortedIter | Known | Known | Known | Depends
QueryIter | Known | Known | Depends | Known 
QueryParIter | Known | Known | Depends | n/a?
QueryParManyIter | Unknown | Unknown | Unknown | n/a?
QueryParManyUniqueIter | Known | Unknown | Unknown | n/a?
QueryCombinationsIter | n/a | Known | Depends | n/a
QuerySortedCombinationsIter | n/a | Known | Known | n/a
QueryCombinationsManyIter | n/a | Unknown | Unknown | n/a
QuerySortedCombinationsManyIter | n/a | Known | Known | n/a

*non-archetypal QueryFilters
**dense EntityLocation/Archetype order

Take this as a rough sketch, it is my current understanding of sensible iterator types, given current/potential API. 
It is not an exhaustive or definite list, I am also least familiar with combination/parallel iterators)
Each of these properties can change what `fetch_next`/its equivalent can do/skip, which makes it really difficult to try and deduplicate without introducing unnecessary complexity or perf hits. I'm not gonna say it is impossible, but in the meantime, documenting these properties would likely make more sense imo.

File: crates/bevy_ecs/src/query/iter.rs
Comment (chescock) [reply crates/bevy_ecs/src/query/iter.rs:2199]: It does compile!  But further testing with miri shows that it isn't sound :(.  I assumed that values wrapped in `ManuallyDrop` no longer need to be valid, but it seems that they do.   (But maybe [that will change?](https://github.com/rust-lang/rfcs/blob/master/text/3336-maybe-dangling.md))  

It does work if we use `MaybeUninit`, though!  We can drop its contents early, and rust doesn't require that they remain valid since they may be uninitialized.  It's more awkward to use, and doesn't `impl Ord`, but I got something that works and seems to make miri happy!  

I don't have permission to make a PR to this PR, so I'll push up the code as a PR in my own repo: chescock/bevy#6

Comment (Victoronz) [reply crates/bevy_ecs/src/query/iter.rs:2199]: Hmm, I am still skeptical? The difference between `mem::drop` and `drop_in_place`/`Drop::drop` is that only `mem::drop` discards/deallocates the value. For the other two, a reference will still be valid after having been dropped! (It is Copy, so has no drop glue).
As such, the for loop with `MaybeUninit::assume_init_drop` does not actually do anything.
Whether this is sound or not I honestly don't know, and even if miri doesn't catch something doesn't mean it is sound.
If it were to be, how does the generated code change? By using a wrapper that opts out of validity like this, we might lose out on optimizations just as much vs if we had just `mem::drop`ped them all with a `collect`

Comment (Victoronz) [reply crates/bevy_ecs/src/query/iter.rs:2199]: Note that in the playground, wrapping with `MaybeUninit` also fails: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=28215602287845f4679dc0f9bdabcbcd

Comment (chescock) [reply crates/bevy_ecs/src/query/iter.rs:2199]: It doesn't seem like we'll get it figured out in a comment thread here, and it's tricky enough that it definitely needs more eyes on it, so I'll plan to push it up as a separate PR once this one is merged.  I'm hoping we agree that if we *can* find a way to do this soundly, that it would be a nice improvement to the API!  :)

Comment (Victoronz) [reply crates/bevy_ecs/src/query/iter.rs:2199]: Feel free to open that PR after! If it can be done soundly and is better than just a plain collect, it'd be nice!

**Pull request merged**
